GVRf/Framework/src/org/gearvrf/io/GVRInputManager.java

```diff
@@ -149,36 +149,46 @@ private GVRCursorType getGVRInputDeviceType(InputDevice device) {
                             : GVRCursorType.UNKNOWN);
                 }
                 return GVRCursorType.MOUSE;
-            } else if ((device.getSources()
-                    & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD
-                    || (device.getSources()
-                            & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK) {
+            } else if ((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD
+                    || (sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK) {
                 return GVRCursorType.CONTROLLER;
+            } else if ((sources & InputDevice.SOURCE_KEYBOARD) == InputDevice.SOURCE_KEYBOARD) {
+                // Allow device keyboard to be a gaze controller if enabled
+                return useGazeCursorController
+                        ? GVRCursorType.GAZE
+                        : GVRCursorType.UNKNOWN;
             }
         }
         return GVRCursorType.UNKNOWN;
     }
 
-    // Return the key if there is one else return -1
-    private long getCacheKey(InputDevice device, GVRCursorType cursorType) {
+    // Return the key if there is one else return 0 (note: -1 is the ID of an Android input device)
+    private long getCacheKey(InputDevice device, int deviceId, GVRCursorType cursorType) {
         if (cursorType != GVRCursorType.UNKNOWN
                 && cursorType != GVRCursorType.EXTERNAL) {
             // Sometimes a device shows up using two device ids
             // here we try to show both devices as one using the
             // product and vendor id
-            long key = device.getVendorId() << 32;
-            key = key | device.getProductId();
+            long key = 0;
+            if (device.getProductId() != 0 && device.getVendorId() != 0) {
+                key = device.getVendorId() << 32;
+                key = key | device.getProductId();
+            } else {
+                // If there is no product/vendor ID (non-USB device), use device Id.
+                // The key/deviceId can be -1 for android's default input
+                key = deviceId;
```

Comment:
There might be an issue with this line. We would be adding multiple gaze controllers when ideally we just need one gaze controller. 

Below is a log I extracted:

```java
D/GVRInputManager( 2383): Device Name: Virtual Device Id: -1 Key: -1
D/GVRInputManager( 2383): Add Controller 0 for Key: -1
D/GVRInputManager( 2383): Device Name: es705 input Device Id: 1 Key: 1
D/GVRInputManager( 2383): Add Controller 1 for Key: 1
D/GVRInputManager( 2383): Device Name: certify_hall Device Id: 3 Key: 0
D/GVRInputManager( 2383): Device Name: sec_touchscreen Device Id: 4 Key: 4
D/GVRInputManager( 2383): Add Controller 2 for Key: 4
D/GVRInputManager( 2383): Device Name: hall Device Id: 5 Key: 0
D/GVRInputManager( 2383): Device Name: gpio_keys.30 Device Id: 6 Key: 1
D/GVRInputManager( 2383): Device Name: Headset Device Id: 7 Key: 7
D/GVRInputManager( 2383): Add Controller 3 for Key: 7
D/GVRInputManager( 2383): Device Name: sec_touchkey Device Id: 8 Key: 8
D/GVRInputManager( 2383): Add Controller 4 for Key: 8
D/GVRInputManager( 2383): Device Name: Samsung Game Pad EI-GP20 Device Id: 21 Key: 42216
D/GVRInputManager( 2383): Add Controller 5 for Key: 42216
D/GVRInputManager( 2383): Device Name: Designer Mouse Device Id: 30 Key: 3167
D/GVRInputManager( 2383): Add Controller 6 for Key: 3167
```

If you need to support all these devices then you might have to forward inputs from all the devices to the one gaze controller that is created.